File: app/src/test/java/com/slukhayka/audiobooks/ui/library/ListenComposerTest.kt
package com.slukhayka.audiobooks.ui.library

import com.slukhayka.audiobooks.data.db.AudiobookEntity
import com.slukhayka.audiobooks.data.db.PlaybackProgressEntity
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Pure JVM tests for the wayfinder #62 ListenComposer (no Robolectric). */
class ListenComposerTest {

    // --- fixtures ----------------------------------------------------------

    private fun book(
        id: String,
        title: String,
        author: String = "Автор",
        totalDurationSeconds: Long = 3600L,
        isDownloaded: Boolean = false,
        isFavorite: Boolean = false,
        seriesTitle: String? = null,
        createdAt: Long = 0L
    ) = AudiobookEntity(
        id = id,
        title = title,
        author = author,
        narrator = "Читець",
        description = "",
        coverDrawableRes = 0,
        genre = "",
        sourceUrl = "https://4read.org/$id.html",
        isDownloaded = isDownloaded,
        totalDurationSeconds = totalDurationSeconds
    ).also {
        // ADR-0009: @Ignore projections set in place.
        it.isFavorite = isFavorite
        it.seriesTitle = seriesTitle
        it.createdAt = createdAt
    }

    private fun progress(
        bookId: String,
        position: Long,
        lastListenedAt: Long = 0L
    ) = PlaybackProgressEntity(
        // ADR-0007: progress is Edition-keyed — one row per rendition.
        editionId = "ed-$bookId",
        bookId = bookId,
        currentChapterIndex = 0,
        currentPositionSeconds = position,
        lastListenedAt = lastListenedAt,
        isCompleted = false
    )

    private fun card(
        book: AudiobookEntity,
        progress: PlaybackProgressEntity? = null,
        totalDuration: Long = book.totalDurationSeconds
    ) = LibraryBook(
        book = book,
        progress = progress,
        cumulativePositionSeconds = progress?.currentPositionSeconds ?: 0L,
        totalDurationSeconds = totalDuration
    )

    private fun prefs(
        order: List<ListenComposer.BlockId> = emptyList(),
        hidden: Set<ListenComposer.BlockId> = emptySet(),
        dismissed: Set<String> = emptySet()
    ) = object : ListenPrefs {
        override val order = order
        override val hiddenBlockIds = hidden
        override val dismissedBookIds = dismissed
    }

    private val NOW = 1_700_000_000_000L
    private val DAY = 24 * 3600 * 1000L

    // --- block eligibility and priority -------------------------------------

    @Test
    fun `empty library composes to no blocks`() {
        val blocks = ListenComposer.compose(emptyList(), nextInSeries = null, prefs = prefs(), now = NOW)
        assertTrue(blocks.isEmpty())
    }

    @Test
    fun `hero is the most recently listened unfinished book`() {
        val b1 = card(book("b1", "Перша"), progress("b1", position = 100L, lastListenedAt = NOW - DAY))
        val b2 = card(book("b2", "Друга"), progress("b2", position = 50L, lastListenedAt = NOW))

        val blocks = ListenComposer.compose(listOf(b1, b2), nextInSeries = null, prefs = prefs(), now = NOW)

        assertEquals(ListenComposer.BlockId.HERO, blocks.first().id)
        assertEquals("b2", blocks.first().books.single().book.id)
        assertNull(blocks.first().reason)
    }

    @Test
    fun `completed hero is excluded and the next best becomes hero`() {
        val done = card(book("done", "Завершена"), progress("done", position = 3600L))
        val half = card(book("half", "Половина"), progress("half", position = 100L, lastListenedAt = NOW))

        val blocks = ListenComposer.compose(listOf(done, half), nextInSeries = null, prefs = prefs(), now = NOW)

        val hero = blocks.first { it.id == ListenComposer.BlockId.HERO }
        assertEquals("half", hero.books.single().book.id)
    }

    @Test
    fun `almost-done block lists books at 80 percent with remaining-time reason`() {
        val near = card(book("near", "Майже"), progress("near", position = 3200L)) // 3200/3600 = 89 %
        val far = card(book("far", "Далеко"), progress("far", position = 100L))

        val blocks = ListenComposer.compose(listOf(near, far), nextInSeries = null, prefs = prefs(), now = NOW)

        val block = blocks.first { it.id == ListenComposer.BlockId.ALMOST_DONE }
        assertEquals(listOf("near"), block.books.map { it.book.id })
        assertTrue(block.reason!!.startsWith("До кінця"))
    }

    @Test
    fun `return block lists dormant unfinished books with days-ago reason`() {
        val dormant = card(book("old", "Старе"), progress("old", position = 100L, lastListenedAt = NOW - 20 * DAY))
        val recent = card(book("new", "Свіже"), progress("new", position = 100L, lastListenedAt = NOW - DAY))

        val blocks = ListenComposer.compose(listOf(dormant, recent), nextInSeries = null, prefs = prefs(), now = NOW)

        val block = blocks.first { it.id == ListenComposer.BlockId.RETURN }
        assertEquals(listOf("old"), block.books.map { it.book.id })
        assertTrue(block.reason!!.contains("20"))
    }

    @Test
    fun `next-in-series block renders the passed next volume`() {
        val next = book("next", "Наступна", seriesTitle = "Сага")
        val blocks = ListenComposer.compose(
            emptyList(),
            nextInSeries = next,
            prefs = prefs(),
            now = NOW
        )

        val block = blocks.first { it.id == ListenComposer.BlockId.NEXT_IN_SERIES }
        assertEquals("next", block.books.single().book.id)
        assertTrue(block.reason!!.contains("Сага"))
    }

    @Test
    fun `travel block lists downloaded books with offline reason`() {
        val offline = card(book("off", "Офлайн", isDownloaded = true))
        val online = card(book("on", "Онлайн"))

        val blocks = ListenComposer.compose(listOf(offline, online), nextInSeries = null, prefs = prefs(), now = NOW)

        val block = blocks.first { it.id == ListenComposer.BlockId.TRAVEL }
        assertEquals(listOf("off"), block.books.map { it.book.id })
        assertTrue(block.reason!!.contains("офлайн"))
    }

    @Test
    fun `short block lists books up to three hours with duration reason`() {
        val short = card(book("s", "Коротка", totalDurationSeconds = 2 * 3600L))
        val long = card(book("l", "Довга", totalDurationSeconds = 10 * 3600L))

        val blocks = ListenComposer.compose(listOf(short, long), nextInSeries = null, prefs = prefs(), now = NOW)

        val block = blocks.first { it.id == ListenComposer.BlockId.SHORT }
        assertEquals(listOf("s"), block.books.map { it.book.id })
        assertTrue(block.reason!!.contains("год"))
    }

    @Test
    fun `favorites block lists favourite books`() {
        val fav = card(book("f", "Улюблена", isFavorite = true))
        val plain = card(book("p", "Звичайна"))

        val blocks = ListenComposer.compose(listOf(fav, plain), nextInSeries = null, prefs = prefs(), now = NOW)

        val block = blocks.first { it.id == ListenComposer.BlockId.FAVORITE_AUTHORS }
        assertEquals(listOf("f"), block.books.map { it.book.id })
    }

    @Test
    fun `recently-added block lists books created this week`() {
        val fresh = card(book("fr", "Нове", createdAt = NOW - DAY))
        val ancient = card(book("an", "Старе", createdAt = NOW - 30 * DAY))

        val blocks = ListenComposer.compose(listOf(fresh, ancient), nextInSeries = null, prefs = prefs(), now = NOW)

        val block = blocks.first { it.id == ListenComposer.BlockId.RECENTLY_ADDED }
        assertEquals(listOf("fr"), block.books.map { it.book.id })
        assertTrue(block.reason!!.contains("тижня"))
    }

    // --- default priority ---------------------------------------------------

    @Test
    fun `blocks follow the default product priority`() {
        val all = listOf(
            card(book("a", "Книга", totalDurationSeconds = 10 * 3600L), progress("a", position = 100L, lastListenedAt = NOW)),
            card(book("b", "Майже", totalDurationSeconds = 2 * 3600L), progress("b", position = 6000L, lastListenedAt = NOW - DAY)),
            card(book("c", "Офлайн", isDownloaded = true, totalDurationSeconds = 10 * 3600L)),
            card(book("d", "Улюблена", isFavorite = true, totalDurationSeconds = 10 * 3600L)),
            card(book("e", "Нова", createdAt = NOW, totalDurationSeconds = 10 * 3600L))
        )

        val blocks = ListenComposer.compose(all, nextInSeries = book("n", "Наступна"), prefs = prefs(), now = NOW)

        assertEquals(
            listOf(
                ListenComposer.BlockId.HERO,
                ListenComposer.BlockId.ALMOST_DONE,
                ListenComposer.BlockId.NEXT_IN_SERIES,
                ListenComposer.BlockId.TRAVEL,
                ListenComposer.BlockId.SHORT,
                ListenComposer.BlockId.FAVORITE_AUTHORS,
                ListenComposer.BlockId.RECENTLY_ADDED
            ),
            blocks.map { it.id }
        )
    }

    // --- prefs: order / hidden / dismissed ----------------------------------

    @Test
    fun `user order wins over the default priority`() {
        val all = listOf(
            card(book("a", "Книга", totalDurationSeconds = 10 * 3600L), progress("a", position = 100L, lastListenedAt = NOW)),
            card(book("c", "Офлайн", isDownloaded = true, totalDurationSeconds = 10 * 3600L))
        )

        val blocks = ListenComposer.compose(
            all,
            nextInSeries = null,
            prefs = prefs(order = listOf(ListenComposer.BlockId.TRAVEL, ListenComposer.BlockId.HERO)),
            now = NOW
        )

        assertEquals(
            listOf(ListenComposer.BlockId.TRAVEL, ListenComposer.BlockId.HERO),
            blocks.map { it.id }
        )
    }

    @Test
    fun `hidden blocks are still computed by the composer and filtered by the screen`() {
        val all = listOf(
            card(book("a", "Книга", totalDurationSeconds = 10 * 3600L), progress("a", position = 100L, lastListenedAt = NOW)),
            card(book("c", "Офлайн", isDownloaded = true, totalDurationSeconds = 10 * 3600L))
        )

        val blocks = ListenComposer.compose(
            all,
            nextInSeries = null,
            prefs = prefs(hidden = setOf(ListenComposer.BlockId.TRAVEL)),
            now = NOW
        )

        // The composer itself does not filter — hidden stays computed; the UI
        // decides what to render. This keeps re-showing instant, no state loss.
        assertEquals(2, blocks.size)
        assertTrue(blocks.any { it.id == ListenComposer.BlockId.TRAVEL })
    }

    @Test
    fun `dismissed works are filtered from every block`() {
        val near = card(book("near", "Майже"), progress("near", position = 3400L, lastListenedAt = NOW))
        val other = card(book("other", "Інша"), progress("other", position = 3400L, lastListenedAt = NOW - DAY))

        val blocks = ListenComposer.compose(
            listOf(near, other),
            nextInSeries = null,
            prefs = prefs(dismissed = setOf("near")),
            now = NOW
        )

        val block = blocks.first { it.id == ListenComposer.BlockId.ALMOST_DONE }
        assertEquals(listOf("other"), block.books.map { it.book.id })
    }

    @Test
    fun `an unknown order id falls back to the default position`() {
        val all = listOf(
            card(book("a", "Книга", totalDurationSeconds = 10 * 3600L), progress("a", position = 100L, lastListenedAt = NOW))
        )

        val blocks = ListenComposer.compose(
            all,
            nextInSeries = null,
            prefs = prefs(order = listOf(ListenComposer.BlockId.RETURN, ListenComposer.BlockId.HERO)),
            now = NOW
        )

        assertEquals(listOf(ListenComposer.BlockId.HERO), blocks.map { it.id })
    }
}

File: web/src/ui/listenComposer.test.ts
/**
 * #585 W2.1 — the shelf rules engine: each block's eligibility and reason,
 * the prefs' effective order and hiding, and the full cross-shelf dedup
 * (the hero claims first and never leaks).
 */
import { describe, expect, it } from 'vitest'
import type { LibraryBookView } from './libraryModel'
import {
  composeListenBlocks,
  deduplicateListenShelves,
  effectiveBlockOrder,
  formatRemainingTime,
} from './listenComposer'

const view = (mergeKey: string, overrides: Partial<LibraryBookView> = {}): LibraryBookView => ({
  mergeKey,
  title: `Книга ${mergeKey}`,
  author: 'Автор',
  createdAt: 0,
  status: 'new',
  lastListenedAt: null,
  narrator: '',
  ...overrides,
})

const NOW = 1_000_000_000_000
const DAY = 24 * 3600 * 1000

describe('effectiveBlockOrder', () => {
  it('defaults to the product priority', () => {
    expect(effectiveBlockOrder([])).toEqual([
      'hero', 'almost-done', 'return', 'next-in-series', 'travel', 'short', 'favorite-authors', 'recently-added',
    ])
  })

  it('the user reorder wins; unknown ids fall back to the default position', () => {
    expect(effectiveBlockOrder(['short' as never, 'hero' as never, 'bogus' as never]).slice(0, 2)).toEqual(['short', 'hero'])
    const order = effectiveBlockOrder(['short', 'hero'] as never[])
    expect(order).toHaveLength(8)
    expect(new Set(order).size).toBe(8)
  })
})

describe('composeListenBlocks', () => {
  it('a cold-start library composes nothing', () => {
    expect(composeListenBlocks([], { order: [], hidden: [], dismissed: [] }, { now: NOW })).toEqual([])
  })

  it('the hero is the most recently listened, unfinished book', () => {
    const blocks = composeListenBlocks(
      [
        view('a', { status: 'listening', lastListenedAt: NOW - 2 * DAY }),
        view('b', { status: 'listening', lastListenedAt: NOW - DAY }),
      ],
      { order: [], hidden: [], dismissed: [] },
      { now: NOW },
    )
    expect(blocks[0]!.id).toBe('hero')
    expect(blocks[0]!.books[0]!.mergeKey).toBe('b')
  })

  it('almost-done takes ≥ 80 % and states the honest remaining time', () => {
    const blocks = composeListenBlocks(
      [view('a', { status: 'listening', progress: 0.85, totalSeconds: 3600, cumulativeSeconds: 3600 * 0.85 })],
      { order: [], hidden: [], dismissed: [] },
      { now: NOW },
    )
    const almostDone = blocks.find((block) => block.id === 'almost-done')!
    expect(almostDone.reason).toEqual({ key: 'listenAlmostDoneReason', params: { time: '9 хв' } })
  })

  it('return catches a book dormant for 14+ days with the day count', () => {
    const blocks = composeListenBlocks(
      [view('a', { status: 'listening', lastListenedAt: NOW - 20 * DAY })],
      { order: [], hidden: [], dismissed: [] },
      { now: NOW },
    )
    const ret = blocks.find((block) => block.id === 'return')!
    expect(ret.reason).toEqual({ key: 'listenReturnReason', params: { days: '20 днів' } })
  })

  it('short takes ≤ 3 h totals only', () => {
    const blocks = composeListenBlocks(
      [
        view('shorty', { totalSeconds: 2 * 3600 }),
        view('long', { totalSeconds: 5 * 3600 }),
      ],
      { order: [], hidden: [], dismissed: [] },
      { now: NOW },
    )
    const short = blocks.find((block) => block.id === 'short')!
    expect(short.books.map((book) => book.mergeKey)).toEqual(['shorty'])
  })

  it('recently-added takes only the last week', () => {
    const blocks = composeListenBlocks(
      [
        view('fresh', { createdAt: NOW - 2 * DAY }),
        view('old', { createdAt: NOW - 30 * DAY }),
      ],
      { order: [], hidden: [], dismissed: [] },
      { now: NOW },
    )
    const recent = blocks.find((block) => block.id === 'recently-added')!
    expect(recent.books.map((book) => book.mergeKey)).toEqual(['fresh'])
  })

  it('dismissed works are filtered from every block', () => {
    const blocks = composeListenBlocks(
      [view('a', { status: 'listening', lastListenedAt: NOW - DAY })],
      { order: [], hidden: [], dismissed: ['a'] },
      { now: NOW },
    )
    expect(blocks).toEqual([])
  })

  it('a hidden block stays computed but unrendered', () => {
    const blocks = composeListenBlocks(
      [view('a', { status: 'listening', lastListenedAt: NOW - DAY })],
      { order: [], hidden: ['hero'], dismissed: [] },
      { now: NOW },
    )
    expect(blocks.find((block) => block.id === 'hero')).toBeUndefined()
  })

  it('the user reorder moves an eligible block up', () => {
    const library = [
      view('a', { status: 'listening', lastListenedAt: NOW - DAY }),
      view('b', { createdAt: NOW - DAY }),
    ]
    const blocks = composeListenBlocks(library, { order: ['recently-added', 'hero'], hidden: [], dismissed: [] }, { now: NOW })
    expect(blocks.map((block) => block.id)).toEqual(['recently-added', 'hero'])
  })

  it('the en locale renders the reason params honestly', () => {
    const blocks = composeListenBlocks(
      [view('a', { status: 'listening', lastListenedAt: NOW - 20 * DAY })],
      { order: [], hidden: [], dismissed: [] },
      { now: NOW, locale: 'en' },
    )
    expect(blocks.find((block) => block.id === 'return')!.reason!.params!.days).toBe('20 days')
  })
})

describe('deduplicateListenShelves', () => {
  it('the hero claims its book first; no shelf below repeats it', () => {
    const shared = view('shared', { status: 'listening', lastListenedAt: NOW - DAY, progress: 0.9 })
    const other = view('other', { createdAt: NOW - DAY })
    const blocks = deduplicateListenShelves([
      { id: 'hero', titleKey: 'listenHeroTitle', books: [shared] },
      { id: 'almost-done', titleKey: 'listenAlmostDoneTitle', books: [shared, other] },
    ])
    expect(blocks[0]!.books).toHaveLength(1)
    expect(blocks[1]!.books.map((book) => book.mergeKey)).toEqual(['other'])
  })

  it('reordering a shelf re-prioritises which shelf claims a book', () => {
    const book = view('x')
    const first = deduplicateListenShelves([
      { id: 'short', titleKey: 'listenShortTitle', books: [book] },
      { id: 'recently-added', titleKey: 'listenRecentlyAddedTitle', books: [book] },
    ])
    expect(first[0]!.books).toHaveLength(1)
    expect(first[1]!.books).toHaveLength(0)
    const second = deduplicateListenShelves([
      { id: 'recently-added', titleKey: 'listenRecentlyAddedTitle', books: [book] },
      { id: 'short', titleKey: 'listenShortTitle', books: [book] },
    ])
    expect(second[0]!.books).toHaveLength(1)
    expect(second[1]!.books).toHaveLength(0)
  })

  it('blocks without dupes pass through untouched', () => {
    const blocks = [{ id: 'short' as const, titleKey: 'listenShortTitle' as const, books: [view('a'), view('b')] }]
    expect(deduplicateListenShelves(blocks)).toEqual(blocks)
  })
})

describe('formatRemainingTime', () => {
  it('renders honest uk/en units', () => {
    expect(formatRemainingTime(0)).toBe('0 хв')
    expect(formatRemainingTime(600)).toBe('10 хв')
    expect(formatRemainingTime(3600)).toBe('1 год')
    expect(formatRemainingTime(4800)).toBe('1 год 20 хв')
    expect(formatRemainingTime(4800, 'en')).toBe('1 h 20 min')
  })
})
